### PR TITLE
[SPARK-19965][SS] DataFrame batch reader may fail to infer partitions when reading FileStreamSink's output

### PR DESCRIPTION
**Pull request by lw-lin**
## The Problem

Right now DataFrame batch reader may fail to infer partitions when reading FileStreamSink's output:

```
[info] - partitioned writing and batch reading with 'basePath' *** FAILED *** (3 seconds, 928 milliseconds)
[info]   java.lang.AssertionError: assertion failed: Conflicting directory structures detected. Suspicious paths:
[info] 	***/stream.output-65e3fa45-595a-4d29-b3df-4c001e321637
[info] 	***/stream.output-65e3fa45-595a-4d29-b3df-4c001e321637/_spark_metadata
[info] 
[info] If provided paths are partition directories, please set "basePath" in the options of the data source to specify the root directory of the table. If there are multiple root directories, please load them separately and then union them.
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at org.apache.spark.sql.execution.datasources.PartitioningUtils$.parsePartitions(PartitioningUtils.scala:133)
[info]   at org.apache.spark.sql.execution.datasources.PartitioningUtils$.parsePartitions(PartitioningUtils.scala:98)
[info]   at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex.inferPartitioning(PartitioningAwareFileIndex.scala:156)
[info]   at org.apache.spark.sql.execution.datasources.InMemoryFileIndex.partitionSpec(InMemoryFileIndex.scala:54)
[info]   at org.apache.spark.sql.execution.datasources.PartitioningAwareFileIndex.partitionSchema(PartitioningAwareFileIndex.scala:55)
[info]   at org.apache.spark.sql.execution.datasources.DataSource.getOrInferFileFormatSchema(DataSource.scala:133)
[info]   at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:361)
[info]   at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:160)
[info]   at org.apache.spark.sql.DataFrameReader.parquet(DataFrameReader.scala:536)
[info]   at org.apache.spark.sql.DataFrameReader.parquet(DataFrameReader.scala:520)
[info]   at org.apache.spark.sql.streaming.FileStreamSinkSuite$$anonfun$8.apply$mcV$sp(FileStreamSinkSuite.scala:292)
[info]   at org.apache.spark.sql.streaming.FileStreamSinkSuite$$anonfun$8.apply(FileStreamSinkSuite.scala:268)
[info]   at org.apache.spark.sql.streaming.FileStreamSinkSuite$$anonfun$8.apply(FileStreamSinkSuite.scala:268)
```

## What changes were proposed in this pull request?

This patch alters `InMemoryFileIndex` to filter out these `basePath`s whose ancestor is the streaming metadata dir (`_spark_metadata`). E.g., the following and other similar dir or files will be filtered out:
- (introduced by globbing `basePath/*`)
   - `basePath/_spark_metadata`
- (introduced by globbing `basePath/*/*`)
   - `basePath/_spark_metadata/0`
   - `basePath/_spark_metadata/1`
   - ... 

## How was this patch tested?

Added unit tests

### CODE REVIEW

**Comment by lw-lin** [comment]
@zsxwing would you take a look at this? Thanks!

**Comment by lw-lin** [comment]
Jenkins retest this please

**Comment by lw-lin** [comment]
Jenkins retest this please

**Comment by lw-lin** [comment]
Rebased to master to resolve conflicts

**Comment by lw-lin** [comment]
Jenkins retest this please

**Comment by lw-lin** [comment]
Jenkins retest this please

**Comment by lw-lin** [comment]
@zsxwing would you take a look at your convenience? Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala [hunk 1]
```diff
@@ -53,6 +53,29 @@ object FileStreamSink extends Logging {
...
+
+  /**
+   * Returns true if the path is the metadata dir or its ancestor is the metadata dir.
+   * E.g.:
+   *  - ancestorIsMetadataDirectory(/.../_spark_metadata) => true
+   *  - ancestorIsMetadataDirectory(/.../_spark_metadata/0) => true
+   *  - ancestorIsMetadataDirectory(/a/b/c) => false
+   */
+  def ancestorIsMetadataDirectory(path: Path): Boolean = {
+    require(path.isAbsolute, s"$path is required to be absolute")
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala]: I'm wondering if we can call `makeQualified` instead.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala [hunk 2]
```diff
@@ -53,6 +53,29 @@ object FileStreamSink extends Logging {
...
+    var currentPath = path
+    var finished = false
+    while (!finished) {
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala]: How about changing it to `currentPath  != null`? Then you don't need `finished`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala [hunk 3]
```diff
@@ -145,6 +147,41 @@ class FileStreamSinkSuite extends StreamTest {
     }
   }
 
+  test("partitioned writing and batch reading with 'basePath'") {
+    val inputData = MemoryStream[Int]
+    val ds = inputData.toDS()
+
+    val outputDir = Utils.createTempDir(namePrefix = "stream.output").getCanonicalPath
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: nit: use `withTempDir` to create temp dir instead

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala [hunk 4]
```diff
@@ -145,6 +147,41 @@ class FileStreamSinkSuite extends StreamTest {
...
+    val checkpointDir = Utils.createTempDir(namePrefix = "stream.checkpoint").getCanonicalPath
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: nit: same as above

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 5]
```diff
@@ -36,20 +37,27 @@ import org.apache.spark.util.SerializableConfiguration
...
     partitionSchema: Option[StructType],
     fileStatusCache: FileStatusCache = NoopCache)
   extends PartitioningAwareFileIndex(
     sparkSession, parameters, partitionSchema, fileStatusCache) {
 
+  // Filter out streaming metadata dirs or files such as "/.../_spark_metadata" (the metadata dir)
+  // or "/.../_spark_metadata/0" (a file in the metadata dir). `rootPathsSpecified` might contain
+  // such streaming metadata dir or files, e.g. when after globbing "basePath/*" where "basePath"
+  // is the output of a streaming query.
+  override val rootPaths = rootPathsSpecified.filterNot(FileStreamSink.ancestorIsMetadataDirectory)
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Just to confirm one thing: for files in `rootPaths` or their sub dirs, they will be dropped by `InMemoryFileIndex.shouldFilterOut`. Right?

**Review by zsxwing — CHANGES_REQUESTED** [review]
Sorry for the delay. Looks pretty good. Just some nits.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala (hunk 1 above)
Comment (lw-lin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala]: switched to `makeQualified`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala (hunk 2 above)
Comment (lw-lin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala]: fixed. good point!

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala (hunk 3 above)
Comment (lw-lin) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 5 above)
Comment (lw-lin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Yea your are quite correct! They will be filted by `InMemoryFileIndex.shouldFilterOut`.

**Comment by lw-lin** [comment]
Comments have beed addressed -- @zsxwing it'd be great if you could take another look

**Comment by zsxwing** [comment]
LGTM. Thanks! Merging to master and 2.2.

**Comment by lw-lin** [comment]
thank you @zsxwing

**Pull request closed without merging**